src/thunderbots/software/ai/hl/stp/evaluation/indirect_chip.cpp

```diff
@@ -0,0 +1,258 @@
+        filter_open_triangles(allTriangles, all_enemy_positions);
+
+    target_triangles = remove_outofbounds_triangles(world, target_triangles);
+
+    Point ball_position = world.ball().position();
+
+    return indirect_chip_and_chase_target(target_triangles, ball_position);
+}
+
+
+std::optional<Point> Evaluation::indirect_chip_and_chase_target(
+    const std::vector<Triangle>& triangles, Point ball_position)
+{
+    if (!triangles.empty())
+    {
+        std::pair<Triangle, bool> largest_triangle = get_largest_triangle(
+            triangles,
+            Util::DynamicParameters::Indirect_Chip_Evaluation::min_chip_tri_area.value(),
+            Util::DynamicParameters::Indirect_Chip_Evaluation::min_chip_tri_edge_len
+                .value());
+        Triangle t = largest_triangle.first;
+
+        Point target = get_triangle_center_and_area(t).first;
+        target       = target.norm(
+            (target - ball_position).len() *
+            Util::DynamicParameters::Indirect_Chip_Evaluation::chip_cherry_power_downscale
+                .value());
+
+        // Target should never be further away than maximum chip power
+        if ((target - ball_position).len() >
+            Util::DynamicParameters::Indirect_Chip_Evaluation::max_chip_power.value())
+        {
+            target =
+                ball_position +
+                (target - ball_position)
+                    .norm(
+                        Util::DynamicParameters::Indirect_Chip_Evaluation::max_chip_power
+                            .value());
+        }
+
+        return std::optional(target);
+    }
+    else
+    {
+        return std::nullopt;
+    }
+}
+
+std::vector<Triangle> Evaluation::get_all_triangles(const World& world,
+                                                    std::vector<Point> enemy_players)
+{
+    std::vector<Triangle> triangles;
+    std::vector<Point> allPts = enemy_players;
+
+    allPts.push_back(world.field().enemyCornerNeg());
+    allPts.push_back(world.field().enemyCornerPos());
+    allPts.push_back(Point(0, world.field().enemyCornerPos().y()));
+    allPts.push_back(Point(0, world.field().enemyCornerNeg().y()));
+
+    for (unsigned int i = 0; i < allPts.size() - 2; i++)
+    {
+        for (unsigned int j = i + 1; j < allPts.size() - 1; j++)
+        {
+            for (unsigned int k = j + 1; k < allPts.size(); k++)
+            {
+                Point p1   = allPts[i];
+                Point p2   = allPts[j];
+                Point p3   = allPts[k];
+                Triangle t = triangle(p1, p2, p3);
+                triangles.push_back(t);
+            }
+        }
+    }
+
+    return triangles;
+}
+
+std::vector<Triangle> Evaluation::filter_open_triangles(std::vector<Triangle> triangles,
+                                                        std::vector<Point> enemy_players)
+{
+    std::vector<Triangle> filtered_triangles;
+    bool containsEnemy = false;
+
+    for (unsigned int i = 0; i < triangles.size(); i++)
```

Comment:
The more I learn!!! :monkey: 

[My understanding](https://www.cprogramming.com/c++11/c++11-ranged-for-loop.html) is that this means: `for (variableType variable : vector)`
and then inside the `for` loop I can just call the `variable` instead of `vector[0], vector[1], vector[2]`... Exciting! 

Made an attempt at implementing this for the two `for` loops in `findOpenTriangles`.

Reply:
That's totally correct! It's worth noting there that there is a slight difference between using a `&` after the variable or not:
```
nums = {1, 2, 3}
for (int i : nums){
    i += 1
}
// value of nums is still {1, 2, 3}

nums = {1, 2, 3}
for (int& i : nums){
    i += 1
}
// value of nums has been updated to {2, 3, 4}
```

Basically, using the `&` means that you get a _reference_ to the variable, so you're dealing with exactly the same variable you think you are, and not using the `&` means you're dealing with a _copy_ of the variable.

Does that make sense?